modin/core/dataframe/base/dataframe/dataframe.py

```diff
@@ -0,0 +1,541 @@
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+"""
+Module contains class ModinDataframe.
+
+ModinDataframe is a parent abstract class for any dataframe class.
+"""
+
+from abc import ABC, abstractmethod
+from typing import List, Hashable, Optional, Callable, Union, Dict
+from enum import Enum
+
+
+class Axis(Enum):  # noqa: PR01
+    """
+    An enum that represents the axis argument provided to the algebra operators.
+
+    The enum has 3 values - ROW_WISE to represent the row axis, COL_WISE to represent the
+    column axis, and CELL_WISE to represent no axis. ROW_WISE operations iterate over the rows
+    COL_WISE operations over the columns, and CELL_WISE operations over block partitions.
+    """
+
+    ROW_WISE = 0
+    COL_WISE = 1
+    CELL_WISE = None
+
+
+class ModinDataframe(ABC):
+    """
+    An abstract class that represents the Parent class for any Dataframe class.
+
+    This class is intended to specify the behaviors that a Dataframe must implement.
+
+    For more details about how these methods were chosen, please refer to this
+    (http://www.vldb.org/pvldb/vol13/p2033-petersohn.pdf) paper, which specifies
+    a Dataframe algebra that this class exposes.
+    """
+
+    @abstractmethod
+    def mask(
+        self,
+        row_labels: Optional[List[Hashable]] = None,
+        row_positions: Optional[List[int]] = None,
+        col_labels: Optional[List[Hashable]] = None,
+        col_positions: Optional[List[int]] = None,
+    ) -> "ModinDataframe":
+        """
+        Mask rows and columns in the dataframe.
+
+        Allow users to perform selection and projection on the row and column labels (named notation),
+        in addition to the row and column number (positional notation).
+
+        Parameters
+        ----------
+        row_labels : list of hashable, optional
+            The row labels to extract.
+        row_positions : list of int, optional
+            The row positions to extract.
+        col_labels : list of hashable, optional
+            The column labels to extract.
+        col_positions : list of int, optional
+            The column positions to extract.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe from the mask provided.
+
+        Notes
+        -----
+        If both row_labels and row_positions are provided, an exception is raised.
+        The same rule applies for col_labels and col_positions.
```

Comment:
nit: specify that a ValueError is raised. Also, add that an exception is also raised if none of col_labels, col_positions, row_labels, row_positions are provided.